native_client/ctcdecode/ctc_beam_search_decoder.h

```diff
@@ -7,66 +7,67 @@
 #include "scorer.h"
 #include "output.h"
 #include "alphabet.h"
-#include "decoderstate.h"
 
-/* Initialize CTC beam search decoder
+class DecoderState {
+  int abs_time_step_;
+  int space_id_;
+  int blank_id_;
+  size_t beam_size_;
+  double cutoff_prob_;
+  size_t cutoff_top_n_;
 
- * Parameters:
- *     alphabet: The alphabet.
- *     class_dim: Alphabet length (plus 1 for space character).
- *     ext_scorer: External scorer to evaluate a prefix, which consists of
- *                 n-gram language model scoring and word insertion term.
- *                 Default null, decoding the input sample without scorer.
- * Return:
- *     A struct containing prefixes and state variables. 
-*/
-DecoderState* decoder_init(const Alphabet &alphabet,
-                           int class_dim,
-                           Scorer *ext_scorer);
+  Scorer* ext_scorer_; // weak
```

Comment:
shouldn't we use `std::weak_ptr` then ?

Reply:
`std::weak_ptr` is the weak counterpart of `std::shared_ptr`. The corresponding owned pointer for this pointer is a `std::unique_ptr`, which has no weak counterpart. Most people use a bare pointer. There's a proposal to add a `std::observer_ptr`: http://open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3840.pdf